lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup.py

```diff
@@ -157,242 +217,1038 @@
+        results = dict(changed=changed)
+        results['dvswitch'] = self.switch_name
+        results['portgroup'] = self.portgroup_name
+        changed_list = []
 
-        if self.dv_switch is None:
-            self.module.fail_json(msg="A distributed virtual switch with name %s does not exist" % self.module.params['switch_name'])
-        self.dvs_portgroup = find_dvspg_by_name(self.dv_switch, self.module.params['portgroup_name'])
+        pg_spec = vim.dvs.DistributedVirtualPortgroup.ConfigSpec()
+        # Use the same version in the new spec; The version will be increased by one by the API automatically
+        pg_spec.configVersion = self.dvs_portgroup.config.configVersion
+        pg_config = self.dvs_portgroup.config
 
-        if self.dvs_portgroup is None:
-            return 'absent'
+        # Check number of ports (only if option is specified)
+        # NOTE: 'self.portgroup_type' is deprecated. Will be removed in 2.11
+        if self.port_binding == 'ephemeral' or self.portgroup_type == 'ephemeral':
+            results['num_ports'] = 'n/a'
         else:
-            return 'present'
+            results['num_ports'] = self.num_ports
+            if self.num_ports and pg_config.numPorts != self.num_ports:
+                changed = True
+                changed_list.append("num ports")
+                results['num_ports_previous'] = pg_config.description
+                pg_spec.numPorts = self.num_ports
+
+        # Check port binding
+        # NOTE: 'self.portgroup_type' is deprecated. Will be removed in 2.11
+        if self.portgroup_type:
+            results['portgroup_type'] = self.portgroup_type
+            if pg_config.type != self.portgroup_type:
+                changed = True
+                changed_list.append("portgroup type")
+                results['portgroup_type_previous'] = self.portgroup_type
+                pg_spec.type = self.portgroup_type
+        else:
+            results['port_binding'] = self.port_binding
+            if pg_config.type != self.get_api_port_binding_mode(self.port_binding):
+                changed = True
+                changed_list.append("port binding")
+                results['port_binding_previous'] = self.get_api_port_binding_mode(pg_config.type)
+                pg_spec.type = self.get_api_port_binding_mode(self.port_binding)
+
+        # Check port allocation
+        if not self.port_allocation:
+            if self.port_binding == 'dynamic' or self.port_binding == 'ephemeral':
+                results['port_allocation'] = 'n/a'
+            elif self.port_binding == 'static':
+                if pg_config.autoExpand:
+                    results['port_allocation'] = 'elastic'
+                else:
+                    results['port_allocation'] = 'fixed'
+        else:
+            results['port_allocation'] = self.port_allocation
+        if self.port_binding == 'static':
+            # if self.port_allocation is None or self.port_allocation == 'elastic':
```

Comment:
These commented lines should be removed to keep the code clean

Reply:
done